sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala

```diff
@@ -194,4 +194,45 @@ class DataFrameNaFunctionsSuite extends QueryTest with SharedSQLContext {
     assert(out1(4) === Row("Amy", null, null))
     assert(out1(5) === Row(null, null, null))
   }
+
+  test("Spark-12231: dropna with partitionBy and groupBy") {
```

Comment:
It would be better to distill this down to its core and put the tests with the other DataSource tests.  In particular it seems like the actual but occurs when there are predicates that hit both partition and normal columns and nothing to do with `na.drop`.  Is that correct?